README.md
# Datalake FTP

Moves files from FTP drop folders to a Cloud Bucket.

## How it works

In the following FTP folder hierarchy, there are two users **bob** and **alice**. 
Each users have four subfolders `INPUT`, `LANDING`, `ARCHIVE` and `QUARANTINE`.

```plaintext
/ftp/home/
|-- bob/
|   |-- INPUT/
|   |-- LANDING/
|   |-- ARCHIVE/
|   `-- QUARANTINE/
`-- alice/
    |-- INPUT/
    |-- LANDING/
    |-- ARCHIVE/
    `-- QUARANTINE/
```

The goal is to synchronize all files the FTP users put in the `INPUT` folder with a cloud bucket ([AWS S3](https://aws.amazon.com/s3/), [Google Cloud Storage](https://cloud.google.com/storage/) or [Azure BlobStorage](https://azure.microsoft.com/en-us/services/storage/blobs/)).
To ensure a file is completely uploaded, its modification time must be _old enough_, like 3 minutes.
The synchronization process is split in the following steps:

1. Select files ready for synchronization: a file is selected if it is in the `INPUT` folder (or one of its subfolders) and if it is older than 3 minutes (according to the last modification time). 
Selected files are moved to the `LANDING` folder.

2. **Optional** Scan files for malwares and viruses: all files in the `LANDING` folder are scanned with [ClamAV antivirus](https://www.clamav.net/).
Infected files are moved to the `QUARANTINE` folder.

3. Copy files to the cloud bucket: all files in the `LANDING` folder are copied along with their subfolders in a target bucket. Files successfully copied are then moved to the `ARCHIVE` folder.

4. Clean up files after retention period: all files older than 72 hours (according to last modification time) are removed from the folders `ARCHIVE` and `QUARANTINE`
     

## Installation

Install the package using pip

```shell
pip install datalake-ftp
```

The following command executes a synchronization with a provided configuration file

```shell
ftpcloud -c 'path/to/config.yaml'
```

The command can be configured as a Linux service. 
For example using systemd:

```ini
[Unit]
Description=Datalake File Transfer
Requires=network.target
After=network.target

[Service]
Type=simple
ExecStart=/usr/local/bin/ftpcloud --daemon --config /etc/datalake-ftp/config.yml
Restart=always
User=datalake
Group=sftp

[Install]
WantedBy=multi-user.target
```


## Configuration

Configuration file is in YAML format.
the following options are available:


| Option | Description | Default value |
| ------ | ----------- | ------------- |
| `ftp_dir` | The parent folder for all FTP users | "/ftp/home" |
| `drop_folder` | The name for the `INPUT` folder | "INPUT" |
| `deliver_folder` | The name for the `LANDING` folder | "LANDING" |
| `archive_folder` | The name for the `ARCHIVE` folder | "ARCHIVE" |
| `quarantine_folder` | The name for the `QUARANTINE` folder | "QUARANTINE" |
| `move_age_seconds` | The age threshold in seconds for selecting files for synchronization | 180 |
| `archive_retention_hours` | The retention period in hours for archived and quarantined files | 72 |
| `cloud.bucket` | The name of the target bucket | "." |
| `cloud.prefix` | The prefix to append to file names before copying to the bucket. | "" |
| `cloud.provider` | The cloud provider for the bucket (either "aws", "gcp", "azure" or "local")| "local" |
| `cloud.monitoring` | The monitoring configuration | `NoMonitor` |
| `antivirus.enabled` | The flag to enable antivirus scanning | `false` |
| `antivirus.params` | Optional arguments to pass to `clamdscan` command | "" |


CHANGELOG.md
## 1.0.2

- Added quarantine to cleanup step

## 1.0.1

- Fixed monitoring issues

## 1.0.0

Initial release


tests/test_ftp_cloud.py
from os import utime
from os.path import join
import pytest
from datalake_ftp import FTPCloud
from pathlib import Path
from shutil import copytree, rmtree
from tempfile import mkdtemp
from pendulum import now
from uuid import uuid4
from datalake.provider.aws import Storage

S3_BUCKET = "eqlab-datamock-ephemeral"


@pytest.fixture(scope="module")
def test_id():
    return uuid4()


@pytest.fixture(scope="module")
def temp_dir():
    temp_dir = mkdtemp(prefix="datalake-ftp", suffix="_home")
    yield temp_dir
    rmtree(temp_dir)


@pytest.fixture(scope="module")
def ftp_dir(temp_dir):
    copytree("./tests/ftp", temp_dir, dirs_exist_ok=True)

    d10 = now().subtract(seconds=10).int_timestamp
    d30 = now().subtract(seconds=30).int_timestamp
    d80 = now().subtract(seconds=80).int_timestamp
    d600 = now().subtract(seconds=600).int_timestamp
    h80 = now().subtract(hours=80).int_timestamp
    utime(join(temp_dir, "bob/INPUT/data-1.csv"), (d600, d600))
    utime(join(temp_dir, "bob/INPUT/data-2.csv"), (d80, d80))
    utime(join(temp_dir, "bob/INPUT/data-3.csv"), (d30, d30))
    utime(join(temp_dir, "alice/INPUT/global.xml"), (d600, d600))
    utime(join(temp_dir, "alice/INPUT/domain/2049_domain.xml"), (d600, d600))
    utime(join(temp_dir, "alice/INPUT/domain/feed/2049_feed.xml"), (d600, d600))
    utime(join(temp_dir, "alice/INPUT/eicar.com.txt"), (d10, d10))
    utime(join(temp_dir, "alice/INPUT/lorem.txt"), (d10, d10))
    utime(join(temp_dir, "jose/ARCHIVE/oldie.txt"), (h80, h80))

    return temp_dir


@pytest.fixture(scope="module")
def ftp_cloud(ftp_dir, test_id):
    config = {
        "cloud": {
            "bucket": S3_BUCKET,
            "prefix": f"ftp-{test_id}",
            "provider": "aws",
            "monitoring": {
                "class": "NoMonitor",
                "params": {
                    "quiet": False,
                },
            },
        },
        "ftp_dir": ftp_dir,
    }

    return FTPCloud(config)


@pytest.fixture(scope="module")
def ftp_cloudav(ftp_dir, test_id):
    config = {
        "cloud": {
            "bucket": S3_BUCKET,
            "prefix": f"ftp-{test_id}",
            "provider": "aws",
            "monitoring": {
                "class": "NoMonitor",
                "params": {
                    "quiet": False,
                },
            },
        },
        "ftp_dir": ftp_dir,
        "antivirus": {"enabled": True, "params": "--config-file tests/clamd.conf --stream"},
    }

    return FTPCloud(config)


def test_scan_files(ftp_cloud):
    l = [str(i) for i in ftp_cloud.scan_folder("INPUT", 120)]
    assert len(l) == 4
    assert "alice/INPUT/global.xml" in l
    assert "bob/INPUT/data-1.csv" in l
    assert "bob/INPUT/data-2.csv" not in l
    assert "bob/INPUT/data-3.csv" not in l

    l = [str(i) for i in ftp_cloud.scan_folder("INPUT", 60)]
    assert len(l) == 5
    assert "alice/INPUT/global.xml" in l
    assert "bob/INPUT/data-1.csv" in l
    assert "bob/INPUT/data-2.csv" in l
    assert "bob/INPUT/data-3.csv" not in l

    l = [str(i) for i in ftp_cloud.scan_folder("INPUT", 20)]
    assert len(l) == 6
    assert "alice/INPUT/global.xml" in l
    assert "bob/INPUT/data-1.csv" in l
    assert "bob/INPUT/data-2.csv" in l
    assert "bob/INPUT/data-3.csv" in l


def test_prepare(ftp_cloud):
    ftp_dir = Path(ftp_cloud.ftp_dir)
    assert (ftp_dir / Path("alice/INPUT/global.xml")).is_file()
    assert not (ftp_dir / Path("alice/LANDING")).is_dir()
    assert not (ftp_dir / Path("alice/LANDING/global.xml")).is_file()
    ftp_cloud.move_to(Path("alice/INPUT/global.xml"), "LANDING")
    assert not (ftp_dir / Path("alice/INPUT/global.xml")).is_file()
    assert (ftp_dir / Path("alice/LANDING/global.xml")).is_file()
    assert (ftp_dir / Path("alice/INPUT/domain/feed/2049_feed.xml")).is_file()

    assert not (ftp_dir / Path("alice/LANDING/domain")).is_dir()
    assert not (ftp_dir / Path("alice/LANDING/domain/feed")).is_dir()
    assert not (ftp_dir / Path("alice/LANDING/domain/feed/2049_feed.xml")).is_file()
    ftp_cloud.move_to(Path("alice/INPUT/domain/feed/2049_feed.xml"), "LANDING")
    assert not (ftp_dir / Path("alice/INPUT/domain/feed/2049_feed.xml")).is_file()
    assert (ftp_dir / Path("alice/LANDING/domain/feed")).is_dir()
    assert (ftp_dir / Path("alice/LANDING/domain/feed/2049_feed.xml")).is_file()


def test_delta3(ftp_cloud):
    ftp_dir = Path(ftp_cloud.ftp_dir)
    assert (ftp_dir / Path("bob/INPUT/data-1.csv")).is_file()
    assert (ftp_dir / Path("alice/INPUT/domain/2049_domain.xml")).is_file()
    assert not (ftp_dir / Path("bob/LANDING/data-1.csv")).is_file()
    assert not (ftp_dir / Path("alice/LANDING/domain/2049_domain.xml")).is_file()
    ftp_cloud.delta3()
    assert not (ftp_dir / Path("bob/INPUT/data-1.csv")).is_file()
    assert not (ftp_dir / Path("alice/INPUT/domain/2049_domain.xml")).is_file()
    assert (ftp_dir / Path("bob/LANDING/data-1.csv")).is_file()
    assert (ftp_dir / Path("alice/LANDING/domain/2049_domain.xml")).is_file()


def test_lambda1(ftp_cloud, test_id):
    ftp_dir = Path(ftp_cloud.ftp_dir)
    cloud = Storage(S3_BUCKET)
    assert (ftp_dir / Path("bob/LANDING/data-1.csv")).is_file()
    assert (ftp_dir / Path("alice/LANDING/domain/2049_domain.xml")).is_file()
    assert not (ftp_dir / Path("bob/ARCHIVE/data-1.csv")).is_file()
    assert not (ftp_dir / Path("alice/ARCHIVE/domain/2049_domain.xml")).is_file()
    assert not cloud.exists(f"ftp-{test_id}/bob/LANDING/data-1.csv")
    assert not cloud.exists(f"ftp-{test_id}/alice/LANDING/domain/2049_domain.xml")
    ftp_cloud.lambda1()
    assert not (ftp_dir / Path("bob/LANDING/data-1.csv")).is_file()
    assert not (ftp_dir / Path("alice/LANDING/domain/2049_domain.xml")).is_file()
    assert (ftp_dir / Path("bob/ARCHIVE/data-1.csv")).is_file()
    assert (ftp_dir / Path("alice/ARCHIVE/domain/2049_domain.xml")).is_file()
    assert cloud.exists(f"ftp-{test_id}/bob/LANDING/data-1.csv")
    assert cloud.exists(f"ftp-{test_id}/alice/LANDING/domain/2049_domain.xml")


@pytest.mark.clamav
def test_antivirus(ftp_cloudav, test_id, caplog):
    ftp_dir = Path(ftp_cloudav.ftp_dir)
    ftp_cloudav.move_to(Path("alice/INPUT/eicar.com.txt"), "LANDING")
    ftp_cloudav.move_to(Path("alice/INPUT/lorem.txt"), "LANDING")
    cloud = Storage(S3_BUCKET)
    assert (ftp_dir / Path("alice/LANDING/eicar.com.txt")).is_file()
    assert (ftp_dir / Path("alice/LANDING/lorem.txt")).is_file()
    assert not (ftp_dir / Path("alice/QUARANTINE/eicar.com.txt")).is_file()
    assert not (ftp_dir / Path("alice/QUARANTINE/lorem.txt")).is_file()
    assert not cloud.exists(f"ftp-{test_id}/alice/LANDING/eicar.com.txt")
    assert not cloud.exists(f"ftp-{test_id}/alice/LANDING/lorem.txt")
    ftp_cloudav.lambda1()
    assert not (ftp_dir / Path("alice/LANDING/eicar.com.txt")).is_file()
    assert not (ftp_dir / Path("alice/LANDING/lorem.txt")).is_file()
    assert (ftp_dir / Path("alice/QUARANTINE/eicar.com.txt")).is_file()
    assert not (ftp_dir / Path("alice/QUARANTINE/lorem.txt")).is_file()
    assert not cloud.exists(f"ftp-{test_id}/alice/LANDING/eicar.com.txt")
    assert cloud.exists(f"ftp-{test_id}/alice/LANDING/lorem.txt")
    assert caplog.records[0].levelname == "WARNING"
    assert "EICAR" in caplog.text


def test_delta24(ftp_cloud):
    ftp_dir = Path(ftp_cloud.ftp_dir)
    assert (ftp_dir / Path("jose/ARCHIVE/oldie.txt")).is_file()
    assert (ftp_dir / Path("bob/ARCHIVE/data-1.csv")).is_file()
    assert (ftp_dir / Path("alice/ARCHIVE/domain/2049_domain.xml")).is_file()
    ftp_cloud.delta24()
    assert not (ftp_dir / Path("jose/ARCHIVE/oldie.txt")).is_file()
    assert (ftp_dir / Path("bob/ARCHIVE/data-1.csv")).is_file()
    assert (ftp_dir / Path("alice/ARCHIVE/domain/2049_domain.xml")).is_file()


datalake_ftp/__init__.py
from glob import glob
from logging import getLogger
from pathlib import Path
from os import makedirs
from os.path import join
import pendulum
import shutil
from subprocess import run
from datalake import ServiceDiscovery
from datalake.telemetry import Measurement

STATUS_SUCCESS = "Success"
STATUS_ERROR = "Failure"
STATUS_INFECTED = "Infected"


class FTPCloud:
    def __init__(self, config={}):
        self._logger = getLogger(f"{__name__}.{__class__.__name__}")

        if not isinstance(config, dict):  # pragma: no cover
            raise ValueError("FTPCloud configuration must be a dict")

        self._config = {
            "cloud": {
                "bucket": ".",
                "prefix": None,
                "provider": "local",
                "monitoring": {
                    "class": "NoMonitor",
                    "params": {},
                },
            },
            "ftp_dir": "/ftp/home",
            "drop_folder": "INPUT",
            "deliver_folder": "LANDING",
            "archive_folder": "ARCHIVE",
            "quarantine_folder": "QUARANTINE",
            "move_age_seconds": 180,
            "archive_retention_hours": 72,
            "antivirus": {
                "enabled": False,
                "params": "",
            },
        }
        self._config.update(config)

        self._ftp_dir = Path(self._config["ftp_dir"])
        self._services = ServiceDiscovery(**self._config["cloud"])

    @property
    def ftp_dir(self):
        return self._ftp_dir

    def full_path(self, path):
        return self._ftp_dir / path

    def scan_folder(self, folder, min_age=60):
        """
        List files in a folder across FTP users.

        :param subfolder: the subfolder in FTP users' home to scan.
        :param min_age: the mtime threshold in seconds (default is 60).
        :return: The ``list`` of ``Path`` which mtime is older that ``min_age`` seconds.
        """
        rtime = pendulum.now()
        scan_files = self._ftp_dir.glob(f"*/{folder}/**/*")
        scan_result = []
        for scan_file in scan_files:
            if scan_file.is_file():
                relative_file = scan_file.relative_to(self._ftp_dir)
                mtime = pendulum.from_timestamp(scan_file.stat().st_mtime)
                etime = mtime.diff(rtime).in_seconds()
                if etime > min_age:
                    scan_result.append(relative_file)
        return scan_result

    def move_to(self, src_path, folder):
        """
        Move a file from a user subfolder to another subfolder (yet for the same user)

        :param src_path: the ``Path`` of the file to move.
        :param folder: the subfolder to move the file to.
        :return: ``None``
        """
        target_parts = list(src_path.parts)
        target_parts[1] = folder
        target_path = self.full_path(Path(*target_parts))
        target_parent = target_path.parent
        makedirs(str(target_parent), mode=0o755, exist_ok=True)
        source_path = self.full_path(src_path)
        shutil.move(str(source_path), str(target_path))

    def delta3(self):
        files_to_move = self.scan_folder(folder=self._config["drop_folder"], min_age=self._config["move_age_seconds"])
        for file_to_move in files_to_move:
            metric = Measurement("ftpcloud-received")
            metric.add_labels({"file_path": str(file_to_move)})
            metric.add_measure("file_size", self.full_path(file_to_move).stat().st_size)
            try:
                self.move_to(file_to_move, self._config["deliver_folder"])
                metric.add_label("status", STATUS_SUCCESS)
            except Exception as e:  # pragma: no cover
                self._logger.error(f"An error occured whilst moving {str(file_to_move)}: {str(e)}")
                metric.add_label("status", STATUS_ERROR)

            self._services.monitor.safe_push(metric)

    def lambda1(self):
        files_to_move = self.scan_folder(folder=self._config["deliver_folder"], min_age=0)
        for file_to_move in files_to_move:
            metric = Measurement("ftpcloud-delivered")
            metric.add_labels({"file_path": str(file_to_move)})
            metric.add_measure("file_size", self.full_path(file_to_move).stat().st_size)
            file_path = str(self._ftp_dir / file_to_move)
            try:
                is_safe = True
                if self._config["antivirus"]["enabled"]:
                    clamav = run(
                        f"clamdscan --no-summary {self._config['antivirus']['params']} {file_path}",
                        shell=True,
                        text=True,
                        capture_output=True,
                    )
                    status = clamav.stdout.replace(f"{file_path}: ", "")
                    if clamav.returncode == 0:
                        is_safe = True
                    elif clamav.returncode == 1:
                        metric.add_label("virus_name", status)
                        metric.add_label("status", STATUS_INFECTED)
                        self._logger.warning(f"Virus detected in file '{str(file_to_move)}': {status}")
                        is_safe = False
                    else:  # pragma: no cover
                        raise IOError(clamav.stderr)

                if is_safe:
                    bucket = self._services.get_storage(self._config["cloud"]["bucket"])
                    if self._config["cloud"]["prefix"] is None:  # pragma: no cover
                        target_path = str(file_to_move)
                    else:
                        target_path = join(self._config["cloud"]["prefix"], str(file_to_move))
                    bucket.upload(
                        src=str(self.full_path(file_to_move)),
                        dst=target_path,
                        content_type="octet/stream",
                    )
                    self.move_to(file_to_move, self._config["archive_folder"])
                    metric.add_label("status", STATUS_SUCCESS)
                else:
                    self.move_to(file_to_move, self._config["quarantine_folder"])
            except Exception as e:  # pragma: no cover
                self._logger.error(f"An error occured whilst delivering '{str(file_to_move)}': {str(e)}")
                metric.add_label("status", STATUS_ERROR)

            self._services.monitor.safe_push(metric)

    def delta24(self):
        archives_to_move = self.scan_folder(
            folder=self._config["archive_folder"],
            min_age=self._config["archive_retention_hours"] * 60 * 60,
        )
        infects_to_move = self.scan_folder(
            folder=self._config["quarantine_folder"],
            min_age=self._config["archive_retention_hours"] * 60 * 60,
        )
        for file_to_move in archives_to_move + infects_to_move:
            try:
                self._logger.info(f"Purging file {str(file_to_move)}")
                (self._ftp_dir / file_to_move).unlink()
            except Exception as e:  # pragma: no cover
                self._logger.warning(f"An error occured whilst purging {str(file_to_move)}: {str(e)}")
